src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java

```diff
@@ -0,0 +1,440 @@
+            {
+                TxnDataName name = TxnDataName.user(select.parameters.refName);
+                checkNotNull(name, "Assignments must be named");
+                checkTrue(selectNames.add(name), DUPLICATE_TUPLE_NAME_MESSAGE, name.name());
+
+                SelectStatement prepared = select.prepare(bindVariables);
+                checkAtMostOneRowSpecified(prepared, INCOMPLETE_PRIMARY_KEY_LET_MESSAGE, LazyToString.lazy(() -> prepared.asCQL(QueryOptions.DEFAULT, state)));
+
+                NamedSelect namedSelect = new NamedSelect(name, prepared);
+                preparedAssignments.add(namedSelect);
+                refSources.put(name, new SelectReferenceSource(prepared));
+            }
+
+            if (dataReferences != null)
+                for (RowDataReference.Raw reference : dataReferences)
+                    reference.resolveReference(refSources);
+
+            NamedSelect returningSelect = null;
+            if (select != null)
+            {
+                SelectStatement prepared = select.prepare(bindVariables);
+                // TODO: Accord saves the result of this read, so limit to a single row until that is no longer true.
+                checkAtMostOneRowSpecified(prepared, INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE, LazyToString.lazy(() -> prepared.asCQL(QueryOptions.DEFAULT, state)));
+                returningSelect = new NamedSelect(TxnDataName.returning(), prepared);
+            }
+
+            List<RowDataReference> returningReferences = null;
+
+            if (returning != null)
+            {
+                // TODO: Eliminate/modify this check if we allow full tuple selections.
+                returningReferences = returning.stream().peek(raw -> checkTrue(raw.column() != null, SELECT_REFS_NEED_COLUMN_MESSAGE))
+                                                        .map(RowDataReference.Raw::prepareAsReceiver)
+                                                        .collect(Collectors.toList());
+            }
+
+            List<ModificationStatement> preparedUpdates = new ArrayList<>(updates.size());
+            
+            // check for any read-before-write updates
+            for (int i = 0; i < updates.size(); i++)
+            {
+                ModificationStatement.Parsed parsed = updates.get(i);
+
+                ModificationStatement prepared = parsed.prepare(bindVariables);
+                checkFalse(prepared.hasConditions(), NO_CONDITIONS_IN_UPDATES_MESSAGE);
+                checkFalse(prepared.isTimestampSet(), NO_TIMESTAMPS_IN_UPDATES_MESSAGE);
+
+                preparedUpdates.add(prepared);
+            }
+
+            List<ConditionStatement> preparedConditions = new ArrayList<>(conditions.size());
+            for (ConditionStatement.Raw condition : conditions)
+                // TODO: Is this synthetic ks name dangerous?
+                preparedConditions.add(condition.prepare("[txn]", bindVariables));
```

Comment:
you need to use the keyspace for the data...  lets say you do the following

```
LET a = (SELECT * FROM ks.table LIMIT 1);
IF ks.fancy(a) THEN
  ...
END
```

since `fancy` belongs to `ks` you need that.  We fixed this bug in 1 other place as well.

Reply:
The grammar doesn't support this at present, and if it did, that's where we would actually get the keyspace name from, correct?